### PR TITLE
Pad when validating addresses

### LINKED ISSUE #4216: Connection Migration Failure when Path MTU is asymmetric
This issue is somewhat similar to the issue with asymmetric path MTUs that @kazuho discovered in #4183.

Consider the following situation: Client and server are communication on one path. The client now probes a new path by sending a PATH_CHALLENGE frame on the new path. The server responds with the PATH_RESPONSE frame, but sends it *on the old path*.
The client now considers the new path validated, and switches over to that path. If the PMTU in the direction from server to client is now smaller than 1200 bytes, it will never receive a packet from the server and the connection will time out. Even worse, if only small packets (for example ACK-only packets) make it through, this might delay the idle timeout indefinitely.

While we could require the client to pad the packet containing the PATH_CHALLENGE to 1200 bytes, this still doesn't help us making sure that the return path also actually supports this packet size.

### PR DESCRIPTION
**Pull request by martinthomson**
This isn't obvious, but if you are validating an address, you are also
validating the path to that address.  If you don't pad these packets,
then you don't know that the path supports a reasonable PMTU.

Require padding of datagrams that carry packets that carry
PATH_CHALLENGE.  This neatly solves the problem.

Closes #4216.

### CODE REVIEW

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -2205,6 +2205,11 @@ connection.
 The endpoint MUST use unpredictable data in every PATH_CHALLENGE frame so that
 it can associate the peer's response with the corresponding PATH_CHALLENGE.
 
+An endpoint MUST expand the payload of datagrams that contain a PATH_CHALLENGE
+frame to at least the smallest allowed maximum packet size of 1200 bytes.
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md:2209]:
```suggestion
frame to at least the smallest allowed maximum datagram size of 1200 bytes.
```

File: draft-ietf-quic-transport.md [hunk 2]
```diff
@@ -2205,6 +2205,11 @@ connection.
 The endpoint MUST use unpredictable data in every PATH_CHALLENGE frame so that
 it can associate the peer's response with the corresponding PATH_CHALLENGE.
 
+An endpoint MUST expand the payload of datagrams that contain a PATH_CHALLENGE
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md:2208]:
```suggestion
An endpoint MUST expand datagrams that contain a PATH_CHALLENGE
```

File: draft-ietf-quic-transport.md [hunk 3]
```diff
@@ -2205,6 +2205,11 @@ connection.
...
+frame to at least the smallest allowed maximum packet size of 1200 bytes.
+Sending UDP datagrams of this size ensures that the network path supports a
+Path Maximum Transmission Unit (PMTU); see {{datagram-size}}.
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md:2211]:
```suggestion
Sending UDP datagrams of this size ensures that the network path from the
endpoint to the peer can be used for QUIC; see {{datagram-size}}.
```

File: draft-ietf-quic-transport.md [hunk 4]
```diff
@@ -2205,6 +2205,18 @@ connection.
...
+
+Note:
+
+: Probing with datagrams of 1200 bytes does not guarantee that the return path
+  also supports the same MTU. Migration to a path that can only carry a 1200
+  byte UDP payload in one direction will result in the peer being unable to
+  validate the path.
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md:2218]:
```suggestion
: Probing with datagrams of 1200 bytes does not guarantee that the return path
  also supports the same datagram size. Migration to a path that can carry a
  1200-byte UDP payload in only one direction will result in the peer being unable to
  validate the return path back to the endpoint.
```

File: draft-ietf-quic-transport.md [hunk 5]
```diff
@@ -368,7 +368,7 @@ The two least significant bits from a stream ID therefore identify a stream as
 one of four types, as summarized in {{stream-id-types}}.
 
 | Bits | Stream Type                      |
-|:-----|:---------------------------------|
+| :--- | :------------------------------- |
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: WTF vscode.

**Review by ianswett — APPROVED** [review]
I'm not sure this completely solves the issue, but it looks sensible.

File: draft-ietf-quic-transport.md [hunk 6]
```diff
@@ -2205,6 +2205,18 @@ connection.
 The endpoint MUST use unpredictable data in every PATH_CHALLENGE frame so that
 it can associate the peer's response with the corresponding PATH_CHALLENGE.
 
+An endpoint MUST expand the payload of datagrams that contain a PATH_CHALLENGE
```
[the remark this answers is not in the document]
Comment (martinthomson) [reply draft-ietf-quic-transport.md]: Yeah, that's true, but is it really necessary to add that detail?

**Pull request merged**